Improve Rank Filter Performance

danieldresser-ie opened the pull request:
Sorry this took so long, probably should have addressed all the exceptional floating point value stuff separately - wasn't expecting that to open as many cans of worms as it did, but I think it's in a pretty good place now.

The interesting bit is the Median filter optimization, which I've tried to document pretty thoroughly.

The performance improvements are pretty good - my benchmarks look like:
```
3 Performance Improvements :

- testPerf (GafferImageTest.DilateTest.DilateTest) : was 5.72s now 0.08s (99% reduction)
- testPerf (GafferImageTest.ErodeTest.ErodeTest) : was 5.57s now 0.09s (98% reduction)
- testPerf (GafferImageTest.MedianTest.MedianTest) : was 7.11s now 0.66s (91% reduction)
```

danieldresser-ie commented on python/GafferImageTest/MergeTest.py:
```diff
@@ -397,6 +397,67 @@ def testModes( self ) :
+		nan = GafferImage.Constant()
+		nan["color"].setValue( imath.Color4f( float( "nan" ) ) )
+
+		# Float plugs by default are capped at 3.4e38 - to get infinity, we multiply that by itself.
+		# Maybe inf should be allowed in FloatPlug - the choice to use std::numeric_limits<T>::max()
+		# may have been made with a focus on ints, without considering that for float, there is a float value
+		# higher than std::numeric_limits<float>::max()
```
Yeah, that's fair enough. I'll just remove the comment. When I wrote it, I don't think I'd thought about how it's pretty easy to just change the cap in places where it really does make sense ( like the output from ImageSampler )

danieldresser-ie commented on src/GafferImage/RankFilter.cpp:
```diff
@@ -181,6 +182,547 @@ Imath::Box2i RankFilter::computeDataWindow( const Gaffer::Context *context, cons
+
+	inline float currentResult()
+	{
+		// Return maximum value of all rows
+		float result = m_values[0];
+		for( int i = 1; i < (int)m_values.size(); i++ )
+		{
+			result = std::max( result, m_values[i] );
+		}
+		return result;
```
I feel a bit conflicted about this ... that does seem to work fine, though for simple code in tight loops I don't really like writing it the wrong way and then relying on the compiler to transform it into what we actually want.

We definitely don't want to be dereferencing a pointer every time we compare to the current value, which I think is what a natural implementation of max_element is doing. It sounds like the optimizer is able to remove that dereference ... kinda awkward to run a disassembler on all this templating, but people talk about this here: https://stackoverflow.com/a/25622769 ... sounds like gcc 9 is able to remove the dereference now, though gcc 4 couldn't ... it's all a bit complicated.

Unless you feel strongly, I think I'm tempted to just leave the code that is clearly doing the right thing.

johnhaddon replied:
I do feel fairly strongly that generally speaking one line calling a `std` function is clearer and less error-prone than 6 lines of custom code. But I didn't know about the dereferencing/optimisation shenanigans discussed in those links, so understand your position. Let's leave as-is.

danieldresser-ie commented on src/GafferImage/RankFilter.cpp:
```diff
@@ -181,6 +182,547 @@ Imath::Box2i RankFilter::computeDataWindow( const Gaffer::Context *context, cons
+		//
+		// This means that the comparison invariant is preserved, but the number of elements in the
+		// lower set may now be wrong - this will be rebalanced when currentResult() is called.
+		int i;
+		for( i = 0; i < m_size.x; i++ )
+		{
+			if( currentRow[i] >= m_splitValue )
+			{
+				break;
+			}
+		}
```
Good point. This was 0.02 second reduction in my test - negligible compared to the original runtime, but about 3% of the new runtime.

I do wonder whether this might not be a win for more realistic sizes of median filter - with a radius less than 10, a linear scan with good branch prediction might beat a more sophisticated approach, but I don't have a test for that currently, so I'll just go with lower_bound for now.

danieldresser-ie commented on src/GafferImage/RankFilter.cpp:
```diff
@@ -181,6 +182,547 @@ Imath::Box2i RankFilter::computeDataWindow( const Gaffer::Context *context, cons
+		std::sort( currentRow, currentRow + m_size.x );
+
+		// Update m_splits for this row to preserve the invariant - it needs to be set so that
+		// currentRow[i] < m_splitValue iff i < m_splits[i]
```
This one was not actually a typo, but it probably doesn't hurt to spell out "if and only if" rather than using the shorthand.

johnhaddon replied:
Every day is a school day - `iff` was a new one to me.

danieldresser-ie commented on python/GafferImageTest/MergeTest.py:
```diff
@@ -723,7 +724,35 @@ def runBoundaryCorrectness( self, scale ):
 
-		self.assertImagesEqual( loop["out"], reader["out"], ignoreMetadata = True, maxDifference = 1e-5 if scale > 1 else 0 )
+		# We should be able to do this simple test, but there's currently a problem: ImageTransform currently
+		# screws up values of inf, even with a simple box filter. I think this is because of the comment in
+		# computeChannelData about the filter support being computed incorrectly. This is very bad for
```
Yeah, clarified that this is about Resample.computeChannelData, and I agree that this should be tackled soon.

johnhaddon replied:
Thanks - I've added "Resample : don't overestimate filter support" to the project board as a reminder.

danieldresser-ie commented:
OK, I think I have now addressed or replied to all comments.

Fortunately the Windows CI failure had a straightforward explanation: I'm using OIIO's noise functions to prepare test data, and it appears that one of those is yielding different results on Windows. There is one test image for large radius median filters that was too slow to regenerate every time with OIIO, so I stored that, but I was generating the source image dynamically using OIIO - the mismatch was between the noise generated source on Linux and Windows. I've worked around this by storing the source image as well ... bit of a shame when the source file is twice as large as the filtered one ( noise doesn't compress well ), but 257K isn't totally out of line with the sizes of other test files, so this is probably a reasonable work around to the OIIO bug.

johnhaddon commented on python/GafferImageTest/SamplerTest.py:
```diff
@@ -92,6 +93,26 @@ def testOutOfBoundsSampleModeBlack( self ) :
+		self.assertEqual( GafferImage.Sampler( inf["out"], "R", dw ).sample( 0, 0 ), float( "inf" ) )
+
+		inf["multiply"].setValue( imath.Color4f( -float( "inf" ) ) )
+		#self.assertEqual( GafferImage.Sampler( inf["out"], "R", dw ).sample( 0, 0 ), -float( "inf" ) )
```
The uncommented test is failing on CI. But looking at it in more detail, I'm confused. The test was added in the same commit as the change to the min/max limits on `ImageSampler.color`, but the test isn't testing ImageSampler at all. Instead it's testing the lower level Sampler class.

> Error: FAIL: testExceptionalValues (GafferImageTest.SamplerTest.SamplerTest)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>  File "/__w/gaffer/gaffer/build/python/GafferImageTest/SamplerTest.py", line 111, in testExceptionalValues
>    self.assertEqual( GafferImage.Sampler( inf["out"], "R", dw ).sample( 0, 0 ), -float( "inf" ) )
> AssertionError: 0.0 != -inf

johnhaddon commented:
> OK, I think I have now addressed or replied to all comments.

Thanks - I've resolved all the discussions except for the uncommented test that is now failing. I think that test perhaps needs rewriting and moving to ImageSamplerTest anyway.

> I've worked around this by storing the source image as well ... bit of a shame when the source file is twice as large as the filtered one ( noise doesn't compress well ), but 257K isn't totally out of line with the sizes of other test files, so this is probably a reasonable work around to the OIIO bug.

I really really want to keep the repo size manageable, and GafferImageTest is currently the biggest problem is this regard. Can we at least use half float? And perhaps we could also reduce the resolution by a half?

danieldresser-ie commented:
Fixed test, added matching ImageSampleTest so we're testing the code that actually changed, converted stored test images to half float. Should be good to go
